### ISSUE TITLE
Support migration of existing campaign tracking data

### ISSUE THREAD

**Issue by kevinoid** [PiwikPRO/plugin-AdvancedCampaignReporting#18]
As discussed in #14, it would be great if the ACR plugin provided a mechanism (manual or automatic) for importing/migrating existing campaign tracking data recorded by the base Piwik installation into ACR.  This would be great for both data continuity and to smooth the installation process for existing users by avoiding confusion about missing data.

**Comment by quba** [comment]
@PiwikPRO/product - would it be possible to implement it?

**Comment by Tobias-Conrad** [comment]
just do it.

**Comment by mattab** [comment]
I tried to see if this was trivial (ie. could be done in half an hour), and tried this code:

```
--- plugins/AdvancedCampaignReporting/API.php	(revision d5f3ae667ff17bf485685e9939bd96424b579a92)
+++ plugins/AdvancedCampaignReporting/API.php	(revision )
@@ -36,6 +36,9 @@
     public function getName($idSite, $period, $date, $segment = false, $expanded = false)
     {
         $dataTable = $this->getDataTable(Archiver::CAMPAIGN_NAME_RECORD_NAME, $idSite, $period, $date, $segment, $expanded);
+
+        $this->mergeStandardCampaignNames($dataTable, $idSite, $period, $date, $segment);
+
         return $dataTable;
     }
 
@@ -79,6 +82,29 @@
     {
         $dataTable = $this->getDataTable(Archiver::HIERARCHICAL_SOURCE_MEDIUM_RECORD_NAME, $idSite, $period, $date, $segment, $expanded = false, $idSubtable);
         return $dataTable;
+    }
+
+    /**
+     * Load "Simple Campaign names" data (ie. data from before AdvancedCampaignReporting was enabled),
+     * and merge it with the new advanced campaign data.
+     *
+     * @param $dataTable
+     * @param $idSite
+     * @param $period
+     * @param $date
+     * @param $segment
+     */
+    private function mergeStandardCampaignNames($dataTable, $idSite, $period, $date, $segment)
+    {
+        $dataTableSimpleCampaign = \Piwik\Plugins\Referrers\API::getInstance()->getCampaigns($idSite, $period, $date, $segment, $expanded = false);
+
+        // We remove sub-tables and only migrate the "Campaign name" information (to KISS / MVP)
+        foreach ($dataTableSimpleCampaign as $row) {
+            $row->removeSubtable();
+        }
+        $dataTable->addDataTable($dataTableSimpleCampaign);
     }
 
 }
\ No newline at end of file
```

But it does not work well (wrong / buggy numbers). Needs further work

**Comment by quba** [comment]
@mattab when would it be possible to tackle this one?

**Comment by mattab** [comment]
@quba it will be done in coming week or so - @diosmosis is on it :+1:

**Comment by mattab** [comment]
fixed in #31  - please update the plugin and the old pre-plugin activation data should now display!

**Comment by kevinoid** [comment]
Looks great.  Thanks @diosmosis and @mattab!
